File: LR1/cpp/src/a/Array.cpp
#include "a/Array.hpp"
#include <iostream>
#include <fstream>
#include <algorithm>

const int MIN_CAPACITY = 10;

void initArray(Array &arr, int initialCapacity) {
    arr.capacity = std::max(initialCapacity, MIN_CAPACITY);
    arr.data = new std::string[arr.capacity];
    arr.size = 0;
}

void ensureCapacity(Array &arr) {
    if (arr.size >= arr.capacity) {
        int newCap = (arr.capacity > 0) ? arr.capacity * 2 : MIN_CAPACITY;
        std::string* newData = new std::string[newCap];
        if (arr.data && arr.size > 0) std::copy(arr.data, arr.data + arr.size, newData);
        delete[] arr.data;
        arr.data = newData;
        arr.capacity = newCap;
    }
}

void shrinkCapacity(Array &arr) {
    if (arr.capacity > MIN_CAPACITY && arr.size < arr.capacity / 4) {
        int newCap = std::max(arr.capacity / 2, MIN_CAPACITY);
        std::string* newData = new std::string[newCap];
        if (arr.data && arr.size > 0) std::copy(arr.data, arr.data + arr.size, newData);
        delete[] arr.data;
        arr.data = newData;
        arr.capacity = newCap;
    }
}

void addEnd(Array &arr, const std::string &val) {
    ensureCapacity(arr);
    arr.data[arr.size++] = val;
}

bool addAt(Array &arr, int index, const std::string &val) {
    if (index < 0 || index > arr.size) return false;
    ensureCapacity(arr);
    for (int i = arr.size; i > index; i--) arr.data[i] = arr.data[i-1];
    arr.data[index] = val;
    arr.size++;
    return true;
}

std::string get(Array &arr, int index) {
    if (index < 0 || index >= arr.size) throw std::out_of_range("Index out of range");
    return arr.data[index];
}

bool set(Array &arr, int index, const std::string &val) {
    if (index < 0 || index >= arr.size) return false;
    arr.data[index] = val;
    return true;
}

bool removeAt(Array &arr, int index) {
    if (index < 0 || index >= arr.size) return false;
    for (int i = index; i < arr.size - 1; i++) arr.data[i] = arr.data[i+1];
    arr.size--;
    shrinkCapacity(arr);
    return true;
}

int length(Array &arr) {
    return arr.size;
}

void readFromFile(Array &arr, const std::string &filename) {
    if (!arr.data || arr.capacity == 0) initArray(arr);
    std::ifstream in(filename);
    if (!in.is_open()) return;
    std::string val;
    while (in >> val) addEnd(arr, val);
    in.close();
}

void writeToFile(Array &arr, const std::string &filename) {
    std::ofstream out(filename);
    for (int i = 0; i < arr.size; i++) {
        out << arr.data[i];
        if (i < arr.size - 1) out << " ";
    }
    out.close();
}

void clearArray(Array &arr) {
    delete[] arr.data;
    arr.data = nullptr;
    arr.size = 0;
    arr.capacity = 0;
}

File: LR1/rust/src/a/Array.rs
use std::cmp::max;
use std::fs::{File, OpenOptions};
use std::io::{self, BufRead, Write};

const MIN_CAPACITY: usize = 10;

pub struct Array {
    pub data: Vec<String>,
    pub capacity: usize,
}

impl Array {
    pub fn new(initial_capacity: usize) -> Self {
        let cap = max(initial_capacity, MIN_CAPACITY);
        Array { data: Vec::with_capacity(cap), capacity: cap }
    }

    fn ensure_capacity(&mut self) {
        if self.data.len() >= self.capacity {
            self.capacity *= 2;
            self.data.reserve(self.capacity - self.data.capacity());
        }
    }

    fn shrink_capacity(&mut self) {
        if self.capacity > MIN_CAPACITY && self.data.len() < self.capacity / 4 {
            let new_cap = max(self.capacity / 2, MIN_CAPACITY);
            self.data.shrink_to(new_cap);
            self.capacity = new_cap;
        }
    }

    pub fn add_end(&mut self, val: &str) {
        self.ensure_capacity();
        self.data.push(val.to_string());
    }

    pub fn add_at(&mut self, index: usize, val: &str) -> bool {
        if index > self.data.len() { return false; }
        self.ensure_capacity();
        self.data.insert(index, val.to_string());
        true
    }

    pub fn get(&self, index: usize) -> Option<&String> {
        self.data.get(index)
    }

    pub fn set(&mut self, index: usize, val: &str) -> bool {
        if index >= self.data.len() { return false; }
        self.data[index] = val.to_string();
        true
    }

    pub fn remove_at(&mut self, index: usize) -> bool {
        if index >= self.data.len() { return false; }
        self.data.remove(index);
        self.shrink_capacity();
        true
    }

    pub fn len(&self) -> usize {
        self.data.len()
    }

    pub fn clear(&mut self) {
        self.data.clear();
        self.capacity = MIN_CAPACITY;
    }

    pub fn read_from_file(&mut self, filename: &str) -> io::Result<()> {
        self.clear();
        let file = File::open(filename)?;
        for line in io::BufReader::new(file).lines() {
            for token in line?.split_whitespace() {
                self.add_end(token);
            }
        }
        Ok(())
    }

    pub fn write_to_file(&self, filename: &str) -> io::Result<()> {
        let mut file = OpenOptions::new().create(true).write(true).truncate(true).open(filename)?;
        for (i, val) in self.data.iter().enumerate() {
            if i > 0 { write!(file, " ")?; }
            write!(file, "{}", val)?;
        }
        Ok(())
    }
}
